Invisibly analytics

Hassija opened the pull request:
<!--
Thank you for your pull request. Please make sure this PR is scoped to one change, and that any added or changed code includes tests with greater than 80% code coverage. See https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#testing-prebidjs for documentation on testing Prebid.js.
-->

## Type of change
<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [ ] Bugfix
- [ ] Feature
- [ ] New bidder adapter  <!--  IMPORTANT: if checking here, also submit your bidder params documentation here https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders --> 
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [x] Other

## Description of change
Updated the logic of triggering prebid events from analytics adapter. Have added an 'pass/fail' switch with 99% fail rate. Events will be fired to back-end api when this switch is in 'pass' state. 
Have updated UT specs for the same. The test-coverage & view-coverage tasks were failing on local after rebasing the fork with latest master. For older version, was able to achieve 93% of coverage. 
```
## Other information
Please update this PR if any issues. The docs repo will be unaffected with this change, let me know if any additional inputs are needed

CC: @jsnellbaker, @mkendall07 & @snapwich .

Hassija commented on modules/invisiblyAnalyticsAdapter.js:
```diff
@@ -55,9 +55,26 @@ let _pageView = {
+    }
+  };
+  // filter only 50% of events
+  let filter = weightedRand({ fail: 0.99, pass: 0.01 }) == 'pass';
```
Forgot to edit comments, we want it reduced to filter 99% cases and not 50%

FilipStamenkovic replied:
Can you, please, then update the comment and the description of this PR?

Hassija replied:
Sure, have pushed commit with updated comments & updated PR description.

Hassija commented on modules/invisiblyAnalyticsAdapter.js:
```diff
@@ -55,9 +55,26 @@ let _pageView = {
+  screenHeight: y,
 };
 
+let weightedFilter = (function () {
```
With just `Math.random() > 0.99`, the accuracy is not that high, so though of looping through the pass/fail specs. 
Let us know your thoughts on this.

FilipStamenkovic replied:
I did some experiments with `Math.random()`:

```
var counter = 0;
for (var i = 0; i < 100000000; i++) {
    counter += Number(Math.random() > 0.99);
}

console.log(counter);`
```

And here are my results:

```
1000560
999359
1000474
1000116
1000180
999773
```

So, I would say that for you case distribution is 'pretty good'.  

And for your looping through the pass/fail specs, in your first iteration of `for` loop you will compare `0.99 <= Math.random()` (because `fail` key will be first in loop). So, bottom line, you are still using `Math.random()` to filter out 99% cases.

Hassija replied:
Have updated the filter logic as per your recommendation.

Hassija commented:
> I added few comments.
> 
> But, looking at the description and the code your plan is not to enable invisibly adapter in 50% of the cases? Are you sure about this, in 50% cases you won't enable your analytics adapter even though user/publisher has explicitly called pbjs to enable your adapter?
> I hope you have this documented somewhere, because I can imagine that users will open issues on github 'complaining' that analytics adapter is not working for them even though they have enabled it?

I agree the need for having this change documented, should we updated the modules\invisiblyAnalyticsAdapter.md file with information? Or do we have any analytics adapter specific documentation available? to avoid the users from opening any issues.

Hassija commented:
> I think updating `.md` file is good option.
> Also, please consider the option to replace immediately invoked function with simple `Math.random() > 0.99`.

Have updated the `.md` file & removed the IFFE block as per your inputs. Let me know if any additional changes are needed.
